pkg/custompluginmonitor/custom_plugin_monitor.go

```diff
@@ -0,0 +1,172 @@
+	"k8s.io/node-problem-detector/pkg/types"
+
+	"github.com/golang/glog"
+	"k8s.io/node-problem-detector/pkg/custompluginmonitor/plugin"
+	"k8s.io/node-problem-detector/pkg/util/tomb"
+)
+
+type customPluginMonitor struct {
+	config     cpmtypes.CustomPluginConfig
+	conditions []types.Condition
+	plugin     *plugin.Plugin
+	resultChan <-chan cpmtypes.Result
+	statusChan chan *types.Status
+	tomb       *tomb.Tomb
+}
+
+// NewCustomPluginMonitorOrDie create a new customPluginMonitor, panic if error occurs.
+func NewCustomPluginMonitorOrDie(configPath string) types.Monitor {
+	l := &customPluginMonitor{
+		tomb: tomb.NewTomb(),
+	}
+	f, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		glog.Fatalf("Failed to read configuration file %q: %v", configPath, err)
+	}
+	err = json.Unmarshal(f, &l.config)
+	if err != nil {
+		glog.Fatalf("Failed to unmarshal configuration file %q: %v", configPath, err)
+	}
+	// Apply configurations
+	err = (&l.config).ApplyConfiguration()
+	if err != nil {
+		glog.Fatalf("Failed to apply configuration for %q. %v", configPath, err)
+	}
+
+	// Validate configurations
+	err = l.config.Validate()
+	if err != nil {
+		glog.Fatalf("Failed to validate custom plugin config %+v. %v", l.config, err)
+	}
+
+	glog.Infof("Finish parsing custom plugin monitor config file: %+v", l.config)
+
+	l.plugin = plugin.NewPlugin(l.config)
+	// A 1000 size channel should be big enough.
+	l.statusChan = make(chan *types.Status, 1000)
+	return l
+}
+
+func (l *customPluginMonitor) Start() (<-chan *types.Status, error) {
+	glog.Info("Start custom plugin monitor")
+	go l.plugin.Run()
+	go l.monitorLoop()
+	return l.statusChan, nil
+}
+
+func (l *customPluginMonitor) Stop() {
+	glog.Info("Stop custom plugin monitor")
+	l.tomb.Stop()
+}
+
+// monitorLoop is the main loop of log monitor.
+func (l *customPluginMonitor) monitorLoop() {
+	l.initializeStatus()
+
+	resultChan := l.plugin.GetResultChan()
+
+	for {
+		select {
+		case result := <-resultChan:
+			glog.V(3).Infof("Receive new plugin result. %+v", result)
+			status := l.generateStatus(result)
+			glog.Infof("New status generated: %+v", status)
+			l.statusChan <- status
+		case <-l.tomb.Stopping():
+			l.plugin.Stop()
+			glog.Infof("Custom plugin monitor stopped")
+			l.tomb.Done()
+			break
+		}
+	}
+}
+
+// generateStatus generates status from the plugin check result.
+func (l *customPluginMonitor) generateStatus(result cpmtypes.Result) *types.Status {
+	timestamp := time.Now()
+	var events []types.Event
+	if result.Rule.Type == types.Temp {
+		// For temporary error only generate event when exit status is above warning
+		if result.ExitStatus >= cpmtypes.NonOK {
+			events = append(events, types.Event{
+				Severity:  types.Warn,
+				Timestamp: timestamp,
+				Reason:    result.Rule.Reason,
+				Message:   result.Message,
+			})
+		}
+	} else {
+		// For permanent error changes the condition
+		for i := range l.conditions {
```

Comment:
Can we also generate an event `XXChanged` here? (Also do that for log monitor).

Then we don't need to add 2 entries for the same script.

Reply:
IMHO, I prefer we separate events and conditions. This will make npd more flexible in case people we only need conditions aside from events.

Reply:
Actually this is an issue I want to fix long time ago. We should always generate an event for condition switch, I think. We also do that in kubelet https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L430.

Usually condition change is not obvious enough to user.